docs/GRAMMAR_UPDATES.md

```diff
@@ -0,0 +1,235 @@
+
+SwitchBlockStatementGroup:
+  SwitchLabel : {SwitchLabel :} BlockStatements
+
+SwitchLabel:
+  case CaseConstant {, CaseConstant}
+  case null [, default]
+  case CasePattern {, CasePattern} [Guard]
+  default
+
+CaseConstant:
+  ConditionalExpression
+
+CasePattern:
+  Pattern
+
+Guard:
+  when Expression
+```
+
+### Identify New Tokens and Update Lexer Grammar
+
+Not every new language feature requires new tokens. But in the case of the `when` expression,
+we need to introduce a new token to represent the `when` keyword. This requires an update
+to the lexer grammar.
+
+```antlr
+LITERAL_WHEN : 'when' ;
+```
+
+Notes:
+
+- We preface literal tokens with `LITERAL_` to distinguish them from other tokens.
+- Lexer rules are written in uppercase by convention.
+
+We will also need to add a
+new [TokenType](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java)
+in this case. However, we may not need to always add a token type for every new lexer token, it
+depends on the use case. An example would be the `LCURLY` (left curly) token, which is used to
+represent the `{`. This token is also a `SLIST` (statement list) in certain contexts. This greatly
+eases static analysis, since we do not need to differentiate between the two tokens by checking
+their context in checks. Token "reuse" is a common pattern, and helps to avoid code duplication and
+having an unnecessarily large number of token types.
+
+### Update Parser Grammar
+
+Next, we need to update the parser grammar to understand in what contexts the new tokens should
+appear. We need to add a new parser rule for the `when` expression. The `when` expression is
+used in the context of a guard in a switch statement or switch expression. We need to update
+the `guard` rule to recognize the `when` expression.
+
+```antlr
+guardedPattern
+    : primaryPattern guard expression
+    ;
+
+// we use the `LAND` token to represent the logical `&&` operator
+guard: ( LAND | LITERAL_WHEN );
+```
+
+Notes:
+
+- Parser rules are written in lower camel case by convention.
+- It is important to consider how many subrules are needed to represent the new language feature.
+  Due to the recursive nature of parsing, Stackoverflow errors can occur if the grammar is too
+  complex with too many subrules.
+- It is good
+  to try to strike a balance between the number of subrules, the complexity of the grammar, and the
+  complexity of JavaAstVisitor to ease maintenance.
+
+### Update JavaAstVisitor
+
+Now, our lexer recognizes the `when` keyword, and our parser recognizes the `when` expression. We
+have updated
+TokenTypes to include `LITERAL_WHEN`. At this point, we are able to parse the new syntax, however,
+the new tokens
+will not appear in the AST unless we update the JavaAstVisitor. This is because our ANTLR grammar
+provides us with a parse tree, which we use the JavaAstVisitor to traverse and build our AST.
+
+```java
```

Comment:
Remove ` java`

Reply:
https://stackoverflow.com/a/75414028/1015848

Reply:
markdownlint fails on this, that's why it is here in the first place

Reply:
Let's disable what is easiest, or put JavaScript instead of java